ckanext/security/anti_csrf.py

```diff
@@ -73,45 +84,155 @@ def get_cookie_token(request):
     If not, an error will occur.
     """
     token = None
-    if request.cookies.has_key(TOKEN_FIELD_NAME):
+    if TOKEN_FIELD_NAME in request.cookies:
         LOG.debug("Obtaining token from cookie")
         token = request.cookies.get(TOKEN_FIELD_NAME)
     if token is None or token.strip() == "":
         csrf_fail("CSRF token is blank")
     return token
 
 
+def _get_user():
+    """ Retrieve the current user object.
+    """
+    return g.userobj
+
+
+def _get_safe_username():
+    """ Retrieve the current username with unsafe characters URL-encoded.
+    """
+    return urllib.quote(_get_user().name, safe='')
+
+
+def validate_token(token):
+    """ Verify the integrity of the provided token.
+    It must have the expected format (hash!timestamp/nonce/username),
+    the hash must match the other values,
+    the username must match the current account,
+    and it must not be older than our limit (currently 30 minutes).
+    TODO Make max age configurable.
```

Comment:
Should this be the same as the session timeout? Or something separate?

Reply:
It shouldn't be any longer than the session timeout, but it doesn't have to be the same. I think there are cases where the session doesn't time out at all, though.